File: heap-sort/heap_sort.py
def heapify(arr, n, i):
    left = 2 * i + 1
    right = 2 * i + 2
    largest = i

    if left < n and arr[left] > arr[i]:
        largest = left

    if right < n and arr[right] > arr[i]:
        largest = right

    if largest != i:
        arr[largest] , arr[i] = arr[i], arr[largest]
    
        heapify(arr, n, largest)


def heap_sort(arr):
    n = len(arr)

    for i in range(n//2, -1, -1):
        heapify(arr, n, i)

    for i in range(n-1, 0, -1):

        arr[i], arr[0] = arr[0], arr[i]

        heapify(arr, i, 0)

if __name__ == "__main__":
    print("heapsort")
    arr = [76,22,43,11,55,100,300,50]
    heap_sort(arr)
    print(arr)

File: heap-sort/heap_sort.go
package main

import "fmt"

func Swap(a *int, b *int) {
	temp := *a
	*a = *b
	*b = temp
}

func Heapify(arr []int, n int, i int) {
	largest := i
	left := int((i * 2) + 1)
	right := int((i * 2) + 2)

	if left <= n && arr[left] > arr[largest] {
		largest = left
	}

	if right <= n && arr[right] > arr[largest] {
		largest = right
	}

	if i != largest {
		Swap(&arr[i], &arr[largest])
		Heapify(arr, n, largest)
	}
}

func main() {
	arr := []int{22, 43, 200, 11, 222, 424, 65, 70}
	n := len(arr) - 1
	for i := int(n / 2); i >= 0; i-- {
		// fmt.Println(i)
		Heapify(arr, n, i)
	}
	for i := n; i > 0; i-- {
		// fmt.Println(i)
		Heapify(arr, i, 0)
		Swap(&arr[0], &arr[i])
	}
	fmt.Print("yola", arr)
}
